File: server/room.py
from typing import Dict, List, overload

from user import User

class Room:
    def __init__(self, id, max_players, min_players):
        self._id = id # incrementing by one for each new room
        self._players = []
        self._max_players = max_players
        self._min_players = min_players
        self._chat_history: List[Dict] = []
        self._admin = None


    def is_free(self):        
        if len(self._players) >= self._max_players:
            free = False
        else:
            free = True
        
        return free


    def get_id(self):
        return self._id


    def append_to_chat_history(self, time, userid, username, message, type="user"):
        
        message_item = {
            "time": time, 
            "userid": userid,
            "username": username,
            "msg": message,
            "type": type
        }

        self._chat_history.append(message_item)

    def add_user_left_message_to_history(self, username, userid):
        self.append_to_chat_history("", userid, username, "", "user-left")
    
    def add_user_joined_message_to_history(self, username, userid):
        self.append_to_chat_history("", userid, username, "", "user-joined")


    def get_chat_history(self):
        return self._chat_history


    def assign_user(self, user):
        self._players.append(user)
        
        if self._admin is None:
            self._admin = user


    def is_user_present(self, userid: str):
        for room_player in self._players:
            if room_player.get_id() == userid:
                return True
        
        return False

    def get_user(self, userid: str):
        for room_player in self._players:
            if room_player.get_id() == userid:
                return room_player
        return None

    def get_player_names(self) -> List[str]:                
        return [player.get_name() for player in self._players]

    def DEBUG_get_player_names_with_id(self): # TODO remove this again
        return [(player.get_name(), player.get_id()) for player in self._players]

    def remove_player(self, userid):
        for i, user in enumerate(self._players):
            if user.get_id() == userid:
                del self._players[i]            

    def is_player_admin(self, user: User):
        if self._admin.get_id() == user.get_id():
            admin_role = True
        else:
            admin_role = False

        return admin_role

    def check_show_instructions_locally(self):
        if len(self._players) >= self._min_players:
            show_instructions = True
        else:
            show_instructions = False
        
        # print('inside check function', show_instructions)
        return show_instructions
    
    def check_show_instructions_globally(self):
        if len(self._players) == self._min_players:
            show_instructions = True
        else:
            show_instructions = False
        
        return show_instructions

File: js-server/room.js
const User = require('./user');

class Room {
    constructor(id, max_players, min_players) {
        this._id = id; // incrementing by one for each new room
        this._players = [];
        this._max_players = max_players;
        this._min_players = min_players;
        this._chat_history = [];
        this._admin = null;
    }

    is_free() {
        return this._players.length < this._max_players;
    }

    get_id() {
        return this._id;
    }

    append_to_chat_history(time, userid, username, message, type = "user") {
        const message_item = {
            "time": time,
            "userid": userid,
            "username": username,
            "msg": message,
            "type": type
        };
        this._chat_history.push(message_item);
    }

    add_user_left_message_to_history(username, userid) {
        this.append_to_chat_history("", userid, username, "", "user-left");
    }

    add_user_joined_message_to_history(username, userid) {
        this.append_to_chat_history("", userid, username, "", "user-joined");
    }

    get_chat_history() {
        return this._chat_history;
    }

    assign_user(user) {
        this._players.push(user);

        if (this._admin === null) {
            this._admin = user;
            user.set_is_admin(true);
        }
    }

    reassign_admin() {
        if (this._players.length > 0) {
            this._players[0].set_is_admin(true);
            this._admin = this._players[0];
        } else {
            this._admin = null;
        }
    }

    is_user_present(userid) {
        return this._players.some(room_player => room_player.get_id() === userid);
    }

    get_user(userid) {
        return this._players.find(room_player => room_player.get_id() === userid) || null;
    }

    get_players() {
        const all_players = this._players.map(player => ({
            'name': player.get_name(),
            'id': player.get_id(),
            'is_admin': player.get_is_admin()
        }));

        return all_players;
    }

    remove_player(userid) {
        this._players = this._players.filter(user => user.get_id() !== userid);

        // no players are left, clean up messages in the room
        if (this._players.length === 0) {
            this._chat_history = [];
        }
    }

    is_player_admin(user) {
        return this._admin && this._admin.get_id() === user.get_id();
    }

    check_show_instructions_locally() {
        return this._players.length >= this._min_players;
    }

    check_show_instructions_globally() {
        return this._players.length === this._min_players;
    }
}

module.exports = Room;